gc_tests/run.rs

```diff
@@ -0,0 +1,256 @@
+    let d = read_to_string(&path).unwrap();
+    let exp = EXPECTED
+        .captures(&d)
+        .expect(&format!(
+            "{} doesn't contain expected test output",
+            test_name
+        ))
+        .get(0)
+        .unwrap()
+        .as_str()
+        .trim();
+    let exptype = match &*exp
+        .lines()
+        .nth(0)
+        .unwrap()
+        .trim_start_matches("///")
+        .trim()
+        .to_lowercase()
+    {
+        "compile-time error:" => ExpType::CompileTimeError,
+        "run-time error:" => ExpType::RunTimeError,
+        "run-time output:" => ExpType::RunTimeOutput,
+        x => panic!("Unknown type '{}'", x)
+    };
+
+    let expbody = exp.lines().skip(1).collect::<Vec<_>>().join("\n");
+    let bindir = env::var(TEST_OBJ_PATH).unwrap();
+    let compile = Command::new("rustc")
+        .args(&["--out-dir", &bindir, path.to_str().unwrap()])
+        .output()
+        .expect("Couldn't run command");
+
+    let runbin = || {
+        let bin = Path::new(&bindir).join(Path::new(path.file_stem().unwrap()));
+        Command::new(bin).output().expect("Couldn't run bin")
+    };
+
+    match exptype {
+        ExpType::CompileTimeError => {
+            if compile.status.success() {
+                output_failure(test_name, compile);
+                false
+            } else {
+                output_success();
+                true
+            }
+        }
+        ExpType::RunTimeError => {
+            if !compile.status.success() {
+                output_failure(test_name, compile);
+                false
+            } else {
+                let output = runbin();
+                if fuzzy_match(&expbody, &output.stderr) {
+                    output_success();
+                    true
+                } else {
+                    output_failure(test_name, output);
+                    false
+                }
+            }
+        }
+        ExpType::RunTimeOutput => {
+            if !compile.status.success() {
+                output_failure(test_name, compile);
+                false
+            } else {
+                let output = runbin();
+                if !output.status.success() {
+                    output_failure(test_name, output);
+                    false
+                } else {
+                    if fuzzy_match(&expbody, &output.stdout) {
+                        output_success();
+                        true
+                    } else {
+                        output_failure(test_name, output);
+                        false
+                    }
+                }
+            }
+        }
+    }
+}
+
+fn main() {
+    let all_tests = find_tests();
+    eprint!("\nrunning {} tests", all_tests.len());
+    let mut any_failed = false;
+
+    // Set up a tmp directory to store our test object binaries and SOs
+    let bindir = env::var(TEST_OBJ_PATH).unwrap();
+    if !Path::new(&bindir).exists() {
+        Command::new("mkdir")
```

Comment:
Is this a temporary dir?

Reply:
Good question. My comment suggests that doesn't it. I forgot to add the teardown code.

Unless you think that this _shouldn't_ be temporary? And instead I should change the comment. Personally I don't think this would be ideal but object files do tend to linger in the `target` dir for files in `examples/` so there is precedent.

Reply:
If the dir is created under `build` or `target` be don't strictly need teardown, but we could...

Have you seen:
https://crates.io/crates/tempdir

Reply:
I think Edd is suggesting using a function which creates a safer test directory?

Reply:
Yep, using tempdir, by defaul the dir would be removed when it goes out of scope. That's not *always* what you want though.

Reply:
Hmm, I setup the directory once in `main` based on some `const` String and then reference it in each test. Using tempdir would mean passing a reference to its name to each test case which might not be ideal.

I'll play around with it and see what is most ergonomic.

Reply:
`std::env::tempdir` might be what you want

Reply:
[Basically you don't want to be in the business of creating temporary directory names: it seems simple, but always goes wrong in the end.]

Reply:
I'm an idiot: `tempdir` isn't enough on its own. But you want something which builds upon that. It's fine if it gives you back a path of some sort that you clean up manually, but you don't want to use a fixed directory name if at all possible.

Reply:
So IIUC, the issue is that we want the temp dir name to be passed in by some environment variable (preferably `TMPDIR` as that's the unixy way) instead of using a hard-coded constant in the test runner?

Reply:
Almost. We want to use an external function which creates the temporary directory or temporary file name for us. It's hard to do this in a safe way, so let's use someone else's already debugged version.

Reply:
Got it (I think). So that we can run the tests like this: ```TMPDIR=`mktemp -d` cargo test```?

Reply:
No ;) We need to use the rust equivalent of `mktemp`. I think it's in a crate these days.

Reply:
Not sure if this helps, but if you use a NamedTemproaryDir, then you can ask for its path.